File: make_double_ratio.py
# import uproot
import numpy as np
import ROOT as R

# Read the ROOT files
file1 = R.TFile.Open("Plot/2018latestgiventau_2.root")
file2 = R.TFile.Open("Plot/2018latesttau_2.root")
output_file = R.TFile("output.root", "RECREATE")
# Extract the histograms
for DM in ["_", "_dm0_", "_dm1_", "_dm1011_"]:
    for WP in ["VVVLoose", "VVLoose", "VLoose", "Loose", "Medium","Tight","VTight","VVTight", ]:
        hist1_mc_num = file1.Get(f"mc_mutau_{WP}{DM}plot_passed") #  file1[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist1_data_num = file1.Get(f"data_mutau_{WP}_{DM}lot_passed") #  file1[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()
        print(hist1_data_num, f"mc_mutau_{WP}{DM}plot_passed")
        hist1_mc_den = file1.Get(f"mc_mutau_{WP}{DM}plot_total") #  file1[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist1_data_den = file1.Get(f"data_mutau_{WP}{DM}plot_total") #  file1[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()

        hist1_mc_num.Divide(hist1_mc_den) #  file1[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist1_data_num.Divide(hist1_data_den) #  file1[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()


        hist2_mc_num = file2.Get(f"mc_mutau_{WP}{DM}plot_passed") #  file2[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist2_data_num = file2.Get(f"data_mutau_{WP}_{DM}lot_passed") #  file2[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()

        hist2_mc_den = file2.Get(f"mc_mutau_{WP}{DM}plot_total") #  file2[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist2_data_den = file2.Get(f"data_mutau_{WP}{DM}plot_total") #  file2[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()

        hist2_mc_num.Divide(hist2_mc_den) #  file1[f"mc_mutau_Medi{DM}m_plot_eff"].to_hist()
        hist2_data_num.Divide(hist2_data_den) #  file1[f"data_mutau_Medi{DM}m_plot_eff"].to_hist()

        # Compute the ratios
        hist1_data_num.Divide(hist1_mc_num)  # hist1_data / hist1_mc
        hist2_data_num.Divide(hist2_mc_num)  # hist2_data / hist2_mc

        # Compute the ratio of the ratios
        hist1_data_num.Divide(hist2_data_num) # ratio1 / ratio2

        print("finished, ", hist1_data_num)
        # Save to a new ROOT file using CERN ROOT
        output_file.cd()
        hist1_data_num.Write(f"{WP}_{DM}lot_doubleratio")
        output_file.Close()

File: make_double_ratio.cpp
#include <TFile.h>
#include <TH1.h>
#include <iostream>

void make_double_ratio() {
    // Read the ROOT files
    TFile* file1 = new TFile("Plot/2018_2p1giventau_2.root", "READ");
    TFile* file2 = new TFile("Plot/2018_2p1tau_2.root", "READ");
    TFile* output_file = new TFile("2018_Factorisation_corrections_DeepTau2017v2p1.root", "RECREATE");

    const char* DMs[] = {"_", "_dm0_", "_dm1_", "_dm1011_"};
    const char* WPs[] = {"VVVLoose", "VVLoose", "VLoose", "Loose", "Medium", "Tight", "VTight", "VVTight"};

    for (const auto& DM : DMs) {
        for (const auto& WP : WPs) {
            // Extract the histograms
            std::cout<< Form("mc_mutau_%s%splot_passed", WP, DM) << std::endl;
            std::cout<<  Form("data_mutau_%s%splot_total", WP, DM)<< std::endl;
            TH1* hist1_mc_num = (TH1*)file1->Get(Form("mc_mutau_%s%splot_passed", WP, DM));
            TH1* hist1_data_num = (TH1*)file1->Get(Form("data_mutau_%s%splot_passed", WP, DM));
            TH1* hist1_mc_den = (TH1*)file1->Get(Form("mc_mutau_%s%splot_total", WP, DM));
            TH1* hist1_data_den = (TH1*)file1->Get(Form("data_mutau_%s%splot_total", WP, DM));

            hist1_mc_num->Divide(hist1_mc_den);
            hist1_data_num->Divide(hist1_data_den);

            TH1* hist2_mc_num = (TH1*)file2->Get(Form("mc_mutau_%s%splot_passed", WP, DM));
            TH1* hist2_data_num = (TH1*)file2->Get(Form("data_mutau_%s%splot_passed", WP, DM));
            TH1* hist2_mc_den = (TH1*)file2->Get(Form("mc_mutau_%s%splot_total", WP, DM));
            TH1* hist2_data_den = (TH1*)file2->Get(Form("data_mutau_%s%splot_total", WP, DM));

            hist2_mc_num->Divide(hist2_mc_den);
            hist2_data_num->Divide(hist2_data_den);

            // Compute the ratios
            hist1_data_num->Divide(hist1_mc_num);
            hist2_data_num->Divide(hist2_mc_num);

            // Compute the ratio of the ratios
            hist1_data_num->Divide(hist2_data_num);

            TF1* fitFunc = new TF1("fitFunc", "pol0", 40, 200);
            hist1_data_num->Fit(fitFunc, "RQ"); // "R" option for the range and "Q" for quiet mode
            // Save to a new ROOT file using CERN ROOT
            output_file->cd();
            hist1_data_num->Write(Form("%s%sdoubleratio", WP, DM));
            fitFunc->Write(Form("%s%sfit", WP, DM));
            std::cout << "Finished: " << hist1_data_num->GetName() << std::endl;

        }
           

    }
    

    // Don't forget to close the output file after writing!


    // Save and close the output file after writing!
    output_file->Close();
}


int main() {
    make_double_ratio();
    return 0;
}
